pkg/metallb/frr.go

```diff
@@ -0,0 +1,675 @@
+
+	return builder
+}
+
+// WithBGPNeighbor defines a single neighbor IP and ASN number.
+func (builder *FrrConfigurationBuilder) WithBGPNeighbor(bgpPeerIP string,
+	remoteAS uint32, routerIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define BGP Neighbor %s with peer IP %s and ASN %v",
+		builder.Definition.Name, bgpPeerIP, remoteAS)
+
+	if net.ParseIP(bgpPeerIP) == nil {
+		glog.V(100).Infof("The peerIP for the frrConfiguration bgp peer contains invalid ip address %s",
+			bgpPeerIP)
+
+		builder.errorMsg = "frrConfiguration 'peerIP' of the BGPPeer contains invalid ip address"
+	}
+
+	if remoteAS < 1 || remoteAS > 65535 {
+		glog.V(100).Infof("The remote ASN of the frrConfiguration must be between 1 and 65535")
+
+		builder.errorMsg = "frrConfiguration 'remote AS' must be between 1 and 65535"
+	}
+
+	// Check if BGP routers exist
+	if len(builder.Definition.Spec.BGP.Routers) == 0 {
+		glog.V(100).Infof("Invalid routerIndex: %d no router was found", routerIndex)
+
+		builder.errorMsg = "No BGP routers defined in WithBGPNeighbor."
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if len(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors) == 0 {
+		builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors = []frrtypes.Neighbor{{
+			Address: bgpPeerIP,
+			ASN:     remoteAS,
+		},
+		}
+	} else {
+		builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors =
+			append(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors,
+				frrtypes.Neighbor{
+					Address: bgpPeerIP,
+					ASN:     remoteAS,
+				})
+	}
+
+	return builder
+}
+
+// WithBGPPassword defines the password used between BGP peers to form adjacency and attaches the password using the
+// neighbor index from the neighbor list.
+func (builder *FrrConfigurationBuilder) WithBGPPassword(bgpPassword string,
+	routerIndex, neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define FRRConfiguration %s in namespace %s with the password: %s",
+		builder.Definition.Name, builder.Definition.Namespace, bgpPassword)
+
+	if bgpPassword == "" {
+		glog.V(100).Infof("The bgpPassword should not be blank")
+
+		builder.errorMsg = fmt.Sprintf("the bgpPassword %s is an empty string", bgpPassword)
+	}
+
+	if strconv.Itoa(neighborIndex) == "" {
+		glog.V(100).Infof("The neighborIndex cannot be blank when adding a BGP password")
+
+		builder.errorMsg = "the 'neighborIndex' cannot be empty when configuring a BGP password"
+	}
+
+	if len(builder.Definition.Spec.BGP.Routers) == 0 {
+		glog.V(100).Infof("Invalid routerIndex: %d no router was found", routerIndex)
+
+		builder.errorMsg = "No BGP routers defined in WithBGPPassword."
+	}
+
+	if neighborIndex < 0 || neighborIndex >= len(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors) {
```

Comment:
better to check that  Nighbor with  builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors[neighborIndex] exists. In that case you don't need any aditional validation for the routerIndex and bgpIndex